src/tools/wasm-link.cc

```diff
@@ -504,14 +504,22 @@ static void write_names_section(Context* ctx) {
   WRITE_UNKNOWN_SIZE(stream);
   write_u32_leb128(stream, total_count, "element count");
   for (const std::unique_ptr<LinkerInputBinary>& binary: ctx->inputs) {
+    size_t delta = 0;
+    uint32_t off;
     for (size_t i = 0; i < binary->debug_names.size(); i++) {
       if (binary->debug_names[i].empty())
         continue;
       if (i < binary->function_imports.size()) {
-        if (!binary->function_imports[i].active)
+        if (!binary->function_imports[i].active) {
+          delta++;
           continue;
+        }
+        off = binary->imported_function_index_offset - delta;
+      } else {
+        off = binary->function_index_offset;
       }
-      write_u32_leb128(stream, i + binary->function_index_offset, "function index");
+
```

Comment:
Hmm, I think are right.  Imported function indexes can't simply be offset by constant.

I guess we should fix that.    If you want to land this in mean time I'd be ok with that (although perhaps we could update the exiting names.txt test rather than adding a new one?)